contracts/staking/EpochSelector.sol

```diff
@@ -0,0 +1,1683 @@
+    }
+
+    /// @notice internal function to recursively search the node
+    /// @param _node address of the node
+    /// @param searchNumber random number used for traversing the tree
+    /// @return Address of the selected node
+    function _weightedSearch(address _node, uint256 searchNumber) public view returns (address) {
+        // |-----------sumOfLeftWeight -------|----balance-----|------sumOfRightWeights------|
+        Node memory node = nodes[_node];
+        (uint256 index1, uint256 index2, uint256 index3) = _getIndexes(_node);
+
+        if (searchNumber <= index1) {
+            return _weightedSearch(node.left, searchNumber);
+        } else if (searchNumber > index1 && searchNumber <= index2) {
+            return _node;
+        } else if (searchNumber > index2 && searchNumber <= index3) {
+            return _weightedSearch(node.right, searchNumber - index2);
+        } else {
+            // _printNode(_node);
+            // console2.log("indexes", index1, index2, index3);
+            // console2.log("search number", searchNumber);
+            revert(ClusterLib.ERROR_OCCURED_DURING_WEIGHTED_SEARCH);
+        }
+    }
+
+    /// @notice Returns the indexes for the node which will be used to traverse the tree
+    /// @param _node Address of the node
+    function _getIndexes(address _node)
+        internal
+        view
+        returns (
+            uint256,
+            uint256,
+            uint256
+        )
+    {
+        Node memory node = nodes[_node];
+        return (
+            node.sumOfLeftBalances,
+            node.sumOfLeftBalances + node.balance,
+            node.sumOfLeftBalances + node.balance + node.sumOfRightBalances
+        );
+    }
+
+    /// @notice Update the balance of the node
+    /// @param node Address of the current node
+    /// @param key Address of the key
+    /// @param diff Difference in the balance of the key
+    function _update(
+        address node,
+        address key,
+        int256 diff
+    ) internal {
+        Node storage currentNode = nodes[node];
+        if (node == key) {
+            diff > 0 ? currentNode.balance += uint256(diff) : currentNode.balance -= uint256(-diff);
+        } else if (key < node) {
+            diff > 0 ? currentNode.sumOfLeftBalances += uint256(diff) : currentNode.sumOfLeftBalances -= uint256(-diff);
+            _update(currentNode.left, key, diff);
+        } else if (key > node) {
+            diff > 0 ? currentNode.sumOfRightBalances += uint256(diff) : currentNode.sumOfRightBalances -= uint256(-diff);
+            _update(currentNode.right, key, diff);
+        } else {
+            revert(ClusterLib.ERROR_OCCURED_DURING_UPDATE);
+        }
+    }
+
+    /// @notice Insert the node to the by searching the position where to add
+    /// @param node Address of the current node
+    /// @param key Address to add
+    /// @param keyBalance Balance of the key
+    function _insert(
+        address node,
```

Comment:
`root` might be a more appropriate name for this as it is the root of the subtree for which you are updating